### PR TITLE
Replace growl errors by modal in VBA flow

### PR DESCRIPTION
**Pull request by aldo-expensify**
Replacing all growls in the VBA flow by the modal.

The error message coming from the backend is being shown in the ConfirmModal now.

Errors messages associated with inputs where we don't handle highlighting yet, like checkboxes, are shown in the ConfirmModal.

Error modal generic message: 'Please double check any highlighted field and try again'

### Fixed Issues

$ https://github.com/Expensify/App/issues/4785

### Tests / QA Steps

Company Step: /bank-account/company

1. Navigate to `/bank-account/company`
2. Fill all inputs, input an invalid 'Zip code' (less than 6 numbers)
3. Click 'Save & continue'
4. Expected results: errors in inputs validated in the front (i.e. zip code, tax id, etc) should be highlighted in the form and a generic message should be shown in a modal.Screen Shot 2021-09-01 at 3 56 23 PM
5. Input a 'Zip Code' of 6 numbers (i.e. `123123`) and input an invalid phone number (i.e. `123`).
6. Click 'Save & continue'
7. Expected results: The phone number is a field validated in the backend, we will display a specific message in the error modal.Screen Shot 2021-09-01 at 3 57 39 PM

ACH Contract / Beneficial Owners Step: /bank-account/contract

1. Navigate to `/bank-account/contract`
2. Check `I accept the terms and conditions` and `I certify that the information provided is true and accurate`
3. Check `Somebody else owns more than 25% of XXX`, this will add a form to input a beneficial owner's information. Don't fill any of the beneficial owner data.
4. Click 'Save & continue'
5. Expected result: Highlighted invalid field (frontend validation) with error modal with generic message Screen Shot 2021-09-01 at 4 20 07 PM

Requestor Step: /bank-account/requestor

1. Navigate to `/bank-account/requestor`
2. Fill all inputs
3. Input a date with invalid format in 'Date of birth', in example: `asd`
4. Expected result: 'Date of birth' field is highlighted with an error, we get the error modal with the generic message (this validation happened in the frontend).Screen Shot 2021-09-01 at 4 09 26 PM
5. Input a date rejected by the backend in the 'Date of birth' , in example: `2020-10-10`
6. Expected result: The error modal displays a specific error about the 'Date of birth' being incorrect. No field is highlighted.Screen Shot 2021-09-01 at 4 12 51 PM

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [ ] iOS
- [x] Android

### CODE REVIEW

File: src/pages/ReimbursementAccount/RequestorStep.js [hunk 1]
```diff
@@ -71,17 +71,18 @@ class RequestorStep extends React.Component {
...
-        if (!isValidIdentity({
+        const identity = {
             street: this.state.requestorAddressStreet,
             state: this.state.requestorAddressState,
             zipCode: this.state.requestorAddressZipCode,
             dob: this.state.dob,
             ssnLast4: this.state.ssnLast4,
-        })) {
+        };
+        if (!isValidIdentity(identity)) {
```
Comment (marcaaron) [inline src/pages/ReimbursementAccount/RequestorStep.js:85]: We can undo this change for now. Let's only make the changes we need to?

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js [hunk 2]
```diff
@@ -238,6 +239,14 @@ class ReimbursementAccountPage extends React.Component {
                             achData={this.props.reimbursementAccount.achData}
                         />
                     )}
+                    <ConfirmModal
+                        title={this.props.translate('companyStep.confirmModalTitle')} // TODO: Update translation key
+                        onConfirm={() => hideErrorModal()}
```
Comment (marcaaron) [inline src/pages/ReimbursementAccount/ReimbursementAccountPage.js:244]: No need to use an arrow function for this can just do `onConfirm={hideErrorModal}`

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js [hunk 3]
```diff
@@ -238,6 +239,14 @@ class ReimbursementAccountPage extends React.Component {
...
+                        prompt={this.props.reimbursementAccount.errorModalMessage || this.props.translate('companyStep.confirmModalPrompt')}
+                        isVisible={Boolean(this.props.reimbursementAccount.isErrorModalVisible)}
```
Comment (marcaaron) [inline src/pages/ReimbursementAccount/ReimbursementAccountPage.js:246]: Any reason to do a `Boolean()` cast here?

File: src/libs/ValidationUtils.js [hunk 4]
```diff
@@ -75,32 +75,38 @@ function isValidSSNLastFour(ssnLast4) {
...
     }
 
     if (!isValidDate(identity.dob)) {
         showBankAccountFormValidationError(translateLocal('bankAccount.error.dob'), true);
+        showErrorModal();
         return false;
     }
 
     if (!isValidSSNLastFour(identity.ssnLast4)) {
         showBankAccountFormValidationError(translateLocal('bankAccount.error.ssnLast4'), true);
```
Comment (marcaaron) [inline src/libs/ValidationUtils.js:101]: We are still passing the `shouldGrowl` argument in all of these.

File: src/libs/actions/BankAccounts.js [hunk 5]
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
 }
 
 /**
- * Set the current error message. Show Growl for errors which are not yet handled by the error Modal.
+ * Show error modal and optionally a specific error message
+ *
+ * @param {String} errorMessage The error message to be displayed in the modal's body. We show translate('companyStep.confirmModalPrompt') if not provided.
+ */
+
+function showErrorModal(errorMessage = null) {
```
Comment (marcaaron) [inline src/libs/actions/BankAccounts.js:611]: Should maybe be more specific for now `showBankAccountErrorModal()` ?

File: src/libs/actions/BankAccounts.js [hunk 6]
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
...
+    Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage: errorMessage});
```
Comment (marcaaron) [inline src/libs/actions/BankAccounts.js:612]: NAB, if you make the argument match the property name we can do this

```js
Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage});
```

File: src/libs/actions/BankAccounts.js [hunk 7]
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
...
+}
+
+/**
+ * Hide error modal
+ */
+function hideErrorModal() {
```
Comment (marcaaron) [inline src/libs/actions/BankAccounts.js:618]: `hideBankAccountErrorModal()`?

**Review by marcaaron — COMMENTED** [review]
Good start. I left a few ideas for improvements.

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js (hunk 3 above)
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/ReimbursementAccountPage.js:246]: The property `this.props.reimbursementAccount.isErrorModalVisible` is undefined at the start, so you get some errors in the console if you pass undefined to `isVisible'

Screen Shot 2021-08-30 at 4 07 43 PM

File: src/libs/ValidationUtils.js (hunk 4 above)
Comment (aldo-expensify) [reply src/libs/ValidationUtils.js:101]: ops, good catch

File: src/libs/actions/BankAccounts.js (hunk 5 above)
Comment (aldo-expensify) [reply src/libs/actions/BankAccounts.js:611]: I have my doubts about that. I feel like the developer can easily know that this function was imported from BankAccount and I like to avoid having to include that in the methods we export, but if you think is better to change it because it can avoid some confusion later.. I can do it anyway

File: src/libs/actions/BankAccounts.js (hunk 6 above)
Comment (aldo-expensify) [reply src/libs/actions/BankAccounts.js]: Done

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js (hunk 2 above)
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/ReimbursementAccountPage.js]: done

File: src/pages/ReimbursementAccount/RequestorStep.js (hunk 1 above)
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/RequestorStep.js]: done

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js (hunk 3 above)
Comment (marcaaron) [reply src/pages/ReimbursementAccount/ReimbursementAccountPage.js:246]: Ah maybe give it a default value? If it's `undefined` then pass `false` ?

File: src/libs/actions/BankAccounts.js (hunk 5 above)
Comment (marcaaron) [reply src/libs/actions/BankAccounts.js]: If you were deciding to use this pattern in other places then you'd have a generic `showErrorModal()` method when in reality it only relates to a single API command. So, I think it's better to be specific, but if you'd prefer to wait for that to become a problem them I'd understand. If you strongly disagree then disagree and we can keep discussing it.

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js (hunk 3 above)
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/ReimbursementAccountPage.js:246]: you mean like `isVisible={this.props.reimbursementAccount.isErrorModalVisible || false}`, or setting the default as default prop in ConfirmModal.js? If it is the second, the linter complains because it doesn't like default props for 'required props'. If it is the first... I guess it is almost the same.

Linter: `propType "isVisible" is required and should not have a defaultProps declaration`

Comment (marcaaron) [reply src/pages/ReimbursementAccount/ReimbursementAccountPage.js:246]: >If it is the first... I guess it is almost the same.

Fair enough, I've seen it both ways, but mostly have seen use of `lodashGet()` when needing a default value.

File: src/libs/actions/BankAccounts.js (hunk 5 above)
Comment (aldo-expensify) [reply src/libs/actions/BankAccounts.js]: I don't strongly disagree, I'll rename it

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js (hunk 3 above)
Comment (aldo-expensify) [reply src/pages/ReimbursementAccount/ReimbursementAccountPage.js:246]: > > If it is the first... I guess it is almost the same.
> 
> Fair enough, I've seen it both ways, but mostly have seen use of `lodashGet()` when needing a default value.

Hmm you are right, lodashGet seems to be used a lot for this in the project... I guess I'll be consistent

File: src/libs/actions/BankAccounts.js [hunk 8]
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
 }
 
 /**
- * Set the current error message. Show Growl for errors which are not yet handled by the error Modal.
+ * Show error modal and optionally a specific error message
+ *
+ * @param {String} errorMessage The error message to be displayed in the modal's body. We show translate('companyStep.confirmModalPrompt') if not provided.
```
[the remark this answers is not in the document]
Comment (aldo-expensify) [reply src/libs/actions/BankAccounts.js]: Removed

File: src/pages/ReimbursementAccount/ReimbursementAccountPage.js (hunk 3 above)
Comment (marcaaron) [reply src/pages/ReimbursementAccount/ReimbursementAccountPage.js:246]: Ok I could call this a NAB (not a blocker) too - seems minor looking again today ;)

**Review by marcaaron — COMMENTED** [review]
Changes look perfect! 🏆 

I think the one thing left to resolve is that there are some places where the confirm modal title does not entirely make sense. Trying to get a reply to the thread here [now](https://expensify.slack.com/archives/C020EPP9B9A/p1630350084180500?thread_ts=1630092331.070800&cid=C020EPP9B9A)

**Comment by aldo-expensify** [comment]
> think the one thing left to resolve is that there are some places where the confirm modal title does not entirely make sense. Trying to get a reply to the thread here [now](https://expensify.slack.com/archives/C020EPP9B9A/p1630350084180500?thread_ts=1630092331.070800&cid=C020EPP9B9A)

To handle this, would it be fine to include an extra parameter on the `showBankAccountErrorModal` like:

```
function showBankAccountErrorModal(errorModalMessage = null, errorModalTitle = null) {
    Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage, errorModalTitle});
}
```
Use it in the display modal like:
```
<ConfirmModal
  title={lodashGet(this.props, 'reimbursementAccount.errorModalTitle', this.props.translate('companyStep.confirmModalTitle'))}
```

Then, when we set the error from the server and want a custom title, we do:

```
showBankAccountErrorModal(error, translateLocal('bankAccount.error.serverErrorModalTitle'));
```

**Comment by marcaaron** [comment]
>To handle this, would it be fine to include an extra parameter on the showBankAccountErrorModal
>Then, when we set the error from the server and want a custom title

Let's wait for feedback first? I haven't seen much discussion about whether these particular errors should have custom titles.

**Comment by marcaaron** [comment]
>Hey @marcaaron, considering we are going to use the Oops header, does the proposed way of handling it above work for you?

Ah no, I think we should make all the headers say "Oops" if that's acceptable? 
I'm not sure there is a need to get very custom on these headers and it is always best to do the simplest thing.

**Comment by aldo-expensify** [comment]
> Ah no, I think we should make all the headers say "Oops" if that's acceptable?
> I'm not sure there is a need to get very custom on these headers and it is always best to do the simplest thing.

Okk, I think I was miss-understanding  that we wanted different headers. I will change the generic header of the error modal to "Oops" then!

**Comment by aldo-expensify** [comment]
Update:
- Replaced error modal header with "Oops" (EN) / "Ups" (ES)
- Moved translation keys related to error modal from `companyStep` into `bankAccount`

**Review by marcaaron — CHANGES_REQUESTED** [review]
Looks perfect there are just two small issues where unnecessary new lines are added. But other than than great work.

File: src/libs/ValidationUtils.js [hunk 9]
```diff
@@ -74,33 +74,39 @@ function isValidSSNLastFour(ssnLast4) {
...
-        showBankAccountFormValidationError(translateLocal('bankAccount.error.ssnLast4'), true);
+        showBankAccountFormValidationError(translateLocal('bankAccount.error.ssnLast4'));
+        showBankAccountErrorModal();
         return false;
     }
 
     return true;
 }
 
+
```
[the remark this answers is not in the document]
Comment (aldo-expensify) [reply src/libs/ValidationUtils.js]: Done

File: src/libs/actions/BankAccounts.js [hunk 10]
```diff
@@ -612,16 +612,29 @@ function validateBankAccount(bankAccountID, validateCode) {
 }
 
 /**
- * Set the current error message. Show Growl for errors which are not yet handled by the error Modal.
+ * Show error modal and optionally a specific error message
+ *
+ * @param {String} errorModalMessage The error message to be displayed in the modal's body.
+ */
+
```
[the remark this answers is not in the document]
Comment (aldo-expensify) [reply src/libs/actions/BankAccounts.js]: Done

**Comment by aldo-expensify** [comment]
Update: removed empty lines

**Review by marcaaron — APPROVED** [review]
Looks good! Thanks for the changes!

Approving the code. But we still need to clean up the PR description because there is nothing listed for QA Steps. We need to put something there so that QA knows what to test (and also figure out what QA is able to test or not). Otherwise, they will inevitably ping you (or me since I reviewed) and come asking if there is QA to do or not :)

Also, we should test this on all platforms to make sure the changes are safe. There's checkboxes listed so that we test on every platform. But if there's a good reason not to then we can mention it somewhere.

**Comment by MitchExpensify** [comment]
Flo is likely offline for the night. Anyone else we can request? @marcaaron

**Comment by aldo-expensify** [comment]
> Looks good! Thanks for the changes!
> 
> Approving the code. But we still need to clean up the PR description because there is nothing listed for QA Steps. We need to put something there so that QA knows what to test (and also figure out what QA is able to test or not). Otherwise, they will inevitably ping you (or me since I reviewed) and come asking if there is QA to do or not :)
> 
> Also, we should test this on all platforms to make sure the changes are safe. There's checkboxes listed so that we test on every platform. But if there's a good reason not to then we can mention it somewhere.

I'll start cleaning this up now (QA steps, outdated pictures), thanks for the reminder!

**Comment by marcaaron** [comment]
>There is one thing in the bank-account/contract step that looks like a bug, but it could also be that it just doesn't work well in the development environment. I'll test it in main branch, it feels unrelated to what we are doing.

Nice! That was just called out by @JmillsExpensify [here](https://expensify.slack.com/archives/C020EPP9B9A/p1630536226414100?thread_ts=1630532268.395300&cid=C020EPP9B9A). I think we'll need to create a follow up issue to look into it.

If anything else pops up that you notice we can check in the free plan channel to see if @MitchExpensify & @kevinksullivan are aware of the issue yet.

**Review by marcaaron — APPROVED** [review]

**Comment by MitchExpensify** [comment]
Nice one @aldo-expensify! 

I'll create a follow up issue to change the error copy we use when the age entered is under our required 18 years old
